flattentool/sort_iati.py

```diff
@@ -0,0 +1,153 @@
+        name_attribute -- the value of the 'name' attribute in the schema, ie.
+                          the name of the element/type etc. being described,
+                          e.g. iati-activities
+        """
+        for tree in self.trees:
+            schema_element = tree.find("xsd:{0}[@name='{1}']".format(tag_name, name_attribute), namespaces=namespaces)
+            if schema_element is not None:
+                return schema_element
+        return schema_element
+
+    def element_loop(self, element, path):
+        """
+        Return information about the children of the supplied element.
+        """
+        a = element.attrib
+        type_elements = []
+        if 'type' in a:
+            complexType = self.get_schema_element('complexType', a['type'])
+            if complexType is not None:
+                type_elements = ( complexType.findall('xsd:choice/xsd:element', namespaces=namespaces) +
+                    complexType.findall('xsd:sequence/xsd:element', namespaces=namespaces) )
+
+        children = ( element.findall('xsd:complexType/xsd:choice/xsd:element', namespaces=namespaces)
+            + element.findall('xsd:complexType/xsd:sequence/xsd:element', namespaces=namespaces)
+            + element.findall("xsd:complexType/xsd:all/xsd:element", namespaces=namespaces)
+            + type_elements)
+        child_tuples = []
+        for child in children:
+            a = child.attrib
+            if 'name' in a:
+                child_tuples.append((a['name'], child, None, a.get('minOccurs'), a.get('maxOccurs')))
+            else:
+                child_tuples.append((a['ref'], None, child, a.get('minOccurs'), a.get('maxOccurs')))
+        return child_tuples
+
+    def create_schema_dict(self, parent_name, parent_element=None):
+        """
+        Created a nested OrderedDict representing the sturucture (and order!) of
+        element in the IATI schema.
+        """
+        if parent_element is None:
+            parent_element = self.get_schema_element('element', parent_name)
+
+        return OrderedDict([(name, self.create_schema_dict(name, element)) for name, element, _, _, _ in self.element_loop(parent_element, '')])
+
+
+def sort_iati_element(element, schema_subdict):
+    """
+    Sort the given elements children according to the order of schema_subdict.
+    """
+    children = list(element)
+    for child in children:
+        element.remove(child)
+    keys = list(schema_subdict.keys())
+
+    def index_key(x):
+        if x.tag in keys:
+            return keys.index(x.tag)
+        else:
+            return len(keys) + 1
+
+    for child in sorted(children, key=index_key):
+        element.append(child)
+        sort_iati_element(child, schema_subdict.get(child.tag, {}))
+
+
+def sort_iati_xml_file(input_file, output_file, schemas):
+    """
+    Sort an IATI XML file according to the schema.
+    """
+    schema_dict = IATISchemaWalker(schemas).create_schema_dict('iati-activity')
```

Comment:
Elsewhere, iati-activity is generalised away as `root_list_path`. So it seems a bit untidy here.

I actually wonder about removing `sort_iati_xml_file` onwards completely, since it’s pretty useless anyway.